subprojects/core/src/main/java/org/gradle/process/internal/DefaultSystemPropertyHandler.java

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.process.internal;
+
+import com.google.common.collect.ImmutableList;
+import org.gradle.api.Named;
+import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.InputDirectory;
+import org.gradle.api.tasks.InputFile;
+import org.gradle.api.tasks.Internal;
+import org.gradle.api.tasks.Nested;
+import org.gradle.api.tasks.OutputDirectory;
+import org.gradle.api.tasks.OutputFile;
+import org.gradle.api.tasks.PathSensitive;
+import org.gradle.api.tasks.PathSensitivity;
+import org.gradle.process.CommandLineArgumentProvider;
+import org.gradle.process.InputOutputValue;
+import org.gradle.process.JavaForkOptions;
+import org.gradle.process.SystemPropertyHandler;
+import org.gradle.util.DeferredUtil;
+
+public class DefaultSystemPropertyHandler implements SystemPropertyHandler {
+    private final JavaForkOptions javaForkOptions;
+
+    public DefaultSystemPropertyHandler(JavaForkOptions javaForkOptions) {
+        this.javaForkOptions = javaForkOptions;
+    }
+
+    @Override
+    public void add(String name, InputOutputValue value) {
+        javaForkOptions.getJvmArgumentProviders().add(new SystemPropertyCommandLineArgument(name, value));
+    }
+
+    @Override
+    public void add(String name, Object value) {
+        javaForkOptions.systemProperty(name, value);
+    }
+
+    @Override
+    public InputOutputValue inputFile(Object file) {
+        return new InputFileValue(file);
+    }
+
+    @Override
+    public InputOutputValue inputDirectory(Object file) {
+        return new InputDirectoryValue(file);
+    }
+
+    @Override
+    public InputOutputValue outputFile(Object file) {
+        return new OutputFileValue(file);
+    }
+
+    @Override
+    public InputOutputValue outputDirectory(Object file) {
+        return new OutputDirectoryValue(file);
+    }
+
+    @Override
+    public InputOutputValue ignored(Object file) {
+        return new IgnoredValue(file);
+    }
```

Comment:
If the object passed to any of these methods is a `Provider`, should we care about retaining its build dependencies?

Reply:
I find it strange if we would say that we ignore the value for snapshotting (not an input) but we care about the value for task dependencies (is an input).

Isn't it also somewhat hard to implement that use case? We would need to have a concept that we care about the dependencies of a property when walking the task graph but that we do not care about its value. I don't think we have something like this currently.

Reply:
I'm still unsure about this...